core/src/main/scala/org/bykn/bosatsu/VarianceFormula.scala

```diff
@@ -0,0 +1,298 @@
+package org.bykn.bosatsu
+
+import cats.Traverse
+import cats.data.State
+import org.bykn.bosatsu.rankn.{DefinedType, Type, TypeEnv}
+import scala.collection.immutable.SortedMap
+
+import cats.implicits._
+
+sealed abstract class VarianceFormula {
+  import VarianceFormula._
+  import Variance._
+
+  def *(that: VarianceFormula): VarianceFormula =
+    (this, that) match {
+      case (Known(a), Known(b)) => Known(a * b)
+      case (p@Known(Phantom), _) => p
+      case (_, p@Known(Phantom)) => p
+      // at least one unknown, and they could be phantom, which acts like a zero
+      case (a, b) => Times(a, b)
+    }
+
+  def +(that: VarianceFormula): VarianceFormula =
+    (this, that) match {
+      case (Known(Phantom), r) => r
+      case (l, Known(Phantom)) => l
+      case (i@Known(Invariant), _) => i
+      case (_, i@Known(Invariant)) => i
+      case (Known(a), Known(b)) => Known(a + b)
+      // at least one unknown, and they could be invariant:
+      case (a, b) => Plus(a, b)
+    }
+
+  def isKnown: Boolean =
+    this match {
+      case Known(_) => true
+      case _ => false
+    }
+
+  def substitute(uk: Unknown, v: Variance): VarianceFormula =
+    this match {
+      case k@Known(_) => k
+      case u0@Unknown(_) => if (u0 == uk) Known(v) else u0
+      case Times(left, right) => left.substitute(uk, v) * right.substitute(uk, v)
+      case Plus(left, right) => left.substitute(uk, v) + right.substitute(uk, v)
+    }
+}
+
+object VarianceFormula {
+  import Variance._
+
+  sealed abstract class Identifier {
+    def toLong: Long
+  }
+
+  object Identifier {
+    private case class IdImpl(toLong: Long) extends Identifier
+    def nextIdent: State[SolutionState, Identifier] =
+      State { ss: SolutionState =>
+        require(ss.next.toLong < Long.MaxValue, "we shouldn't run long enough to overflow Long")
+        (ss.copy(next = IdImpl(ss.next.toLong + 1L)), ss.next)
+      }
+
+    val init: Identifier = IdImpl(0L)
+  }
+
+  case class SolutionState(
+    next: Identifier,
+    constraints: Map[Unknown, VarianceFormula],
+    solutions: Map[Unknown, Variance],
+    unknowns: Set[Unknown]) {
+
+    /**
+     * After we have processed all the constraints, then all the remaining unknowns must be Phantom
+     */
+    def isSolved: Boolean = constraints.isEmpty
+
+    def isUnsolved(u: Unknown): Boolean = constraints.get(u).isDefined
+
+    /**
+     * constrain the left to equal the right
+     */
+    def constrain(left: Unknown, right: VarianceFormula): SolutionState = {
+      val newRight = constraints.get(left) match {
+        case Some(ex) => ex + right
+        case None => right
+      }
+      copy(constraints = constraints.updated(left, newRight))
+    }
+
+    /**
+     * remove any constraints we can solve for. This is the main meat
+     * of the algorithm
+     */
+    def simplify: SolutionState = {
```

Comment:
this method is too large. Let's split out the two strategies (first remove trivial cases, then do iteration).